### PR TITLE
(web) adding CSRF protection to OIDC and SAML callbacks

### LINKED ISSUE #1337: web API should properly handle 404 errors
Currently, we server index.html for all unknown HTTP requests. We should change that to properly handle 404 errors which at the moment we do not.

### LINKED ISSUE #1364: Dynamic node labels are not visible in the Web UI
The web UI is not showing the dynamic labels.
Snippet from a node config:

```
ssh_service:
  enabled: "yes"
  commands:
  - name: arch
    command: [/bin/uname, -v]
    period: 1h0m0s
```

Output of `tsh ls`:

```
Node Name    Node ID                              Address        Labels                                   
------------ ------------------------------------ -------------- ---------------------------------------- 
ev-t1-master 33e52630-c09b-4dd2-97eb-d091c03845c8 127.0.0.1:3022 arch=#1 SMP Debian 4.12.6-1 (2017-08-12) 
```

The web UI says "No assigned labels".

### PR DESCRIPTION
**Pull request by alex-kovoy**
* adds CSRF protection to OIDC and SAML callbacks
* adds additional HTTP header flags for static resources.
* closes #1337 
* closes #1364

### CODE REVIEW

**Comment by alex-kovoy** [comment]
retest this please

File: lib/httplib/csrf/csrf.go [hunk 1]
```diff
@@ -40,70 +40,91 @@ const (
...
-func extractFromRequest(r *http.Request) ([]byte, error) {
-	issued := r.Header.Get(HeaderName)
-	decoded, err := decode(issued)
-	if err != nil {
-		return nil, trace.Wrap(err)
+// compareTokens securely (constant-time) compares request token against the real token
+// from the session.
+func compareTokens(a, b []byte) bool {
+	// this is required as subtle.ConstantTimeCompare does not check for equal
+	// lengths in Go versions prior to 1.3.
```
Comment (klizhentas) [inline lib/httplib/csrf/csrf.go]: we don't support go versions prior to 1.3, so no need to account for that

File: lib/httplib/httplib.go [hunk 2]
```diff
@@ -69,6 +70,19 @@ func MakeStdHandler(fn StdHandlerFunc) http.HandlerFunc {
 	}
 }
 
+// WithCSRFProtection ensures that request to unauthenticated API is checked against CSRF attacks
+func WithCSRFProtection(fn HandlerFunc) httprouter.Handle {
+	hanlderFn := MakeHandler(fn)
+	return func(w http.ResponseWriter, r *http.Request, p httprouter.Params) {
+		err := csrf.VerifyHTTPHeader(r)
+		if err != nil {
+			trace.WriteError(w, trace.AccessDenied("failed to validate CSRF token", err))
```
Comment (klizhentas) [inline lib/httplib/httplib.go]: no need to explain why access is denied, keep it obscure to make it harder for the attacker to guess why the request has failed.

File: lib/services/identity.go [hunk 3]
```diff
@@ -262,9 +262,11 @@ type OIDCAuthRequest struct {
...
 	CheckUser bool `json:"check_user"`
 
-	// StateToken is generated by service and is used to validate
-	// reuqest coming from
-	StateToken string `json:"state_token"`
+	// ID is request ID generated by service and is used to validate request coming from
+	ID string `json:"id"`
+
+	// State is the state of this request which holds information about user session
+	State AuthRequestSessionState `json:"state"`
```
Comment (klizhentas) [inline lib/services/identity.go]: why create a separate struct with one field? simply add `CSRFToken` as a property.

Can you also keep `StateToken` as a property without renaming it, as far as I understand `StateToken` == `ID` in the new notation.

**Review by klizhentas — CHANGES_REQUESTED** [review]

File: lib/web/apiserver.go [hunk 4]
```diff
@@ -547,8 +546,19 @@ func (h *Handler) oidcLoginWeb(w http.ResponseWriter, r *http.Request, p httprou
 	if connectorID == "" {
 		return nil, trace.BadParameter("missing connector_id query parameter")
 	}
+
+	csrfToken, err := csrf.ExtractTokenFromCookie(r)
+	if connectorID == "" {
```
Comment (r0mant) [inline lib/web/apiserver.go]: Seems wrong. Did you mean `err != nil` maybe? Or `csrfToken == ""`? In any case, the error shouldn't be ignored as well.

File: lib/web/saml.go [hunk 5]
```diff
@@ -26,9 +27,20 @@ func (m *Handler) samlSSO(w http.ResponseWriter, r *http.Request, p httprouter.P
 	if connectorID == "" {
 		return nil, trace.BadParameter("missing connector_id query parameter")
 	}
+
+	csrfToken, err := csrf.ExtractTokenFromCookie(r)
+	if connectorID == "" {
```
Comment (r0mant) [inline lib/web/saml.go]: Same here as above.

File: lib/httplib/csrf/csrf.go [hunk 6]
```diff
@@ -40,70 +40,91 @@ const (
...
 			return "", trace.Wrap(err)
 		}
-	} else {
-		encodedToken = hex.EncodeToString(token)
+	}
+	save(token, w)
+	return token, nil
+}
+
+// VerifyHTTPHeader checks if HTTP header value matches with the cookie.
```
Comment (r0mant) [inline lib/httplib/csrf/csrf.go]: matches the cookie (no "with")

File: lib/web/apiserver.go [hunk 7]
```diff
@@ -547,8 +546,19 @@ func (h *Handler) oidcLoginWeb(w http.ResponseWriter, r *http.Request, p httprou
...
+		return nil, trace.BadParameter("unable to find CSRF token")
```
Comment (r0mant) [inline lib/web/apiserver.go]: Say just "access denied"? You can log a warning here with details.

File: lib/web/apiserver.go [hunk 8]
```diff
@@ -608,6 +618,11 @@ func (h *Handler) oidcCallback(w http.ResponseWriter, r *http.Request, p httprou
 	}
 	// if we created web session, set session cookie and redirect to original url
 	if response.Req.CreateWebSession {
+		err = csrf.VerifyToken(response.Req.State.CSRFToken, r)
+		if err != nil {
+			return nil, trace.AccessDenied("unable to verify CSRF token", err)
```
Comment (r0mant) [inline lib/web/apiserver.go]: Same here - "access denied" + warning?

File: lib/web/saml.go [hunk 9]
```diff
@@ -26,9 +27,20 @@ func (m *Handler) samlSSO(w http.ResponseWriter, r *http.Request, p httprouter.P
...
+		return nil, trace.BadParameter("unable to find CSRF token")
```
Comment (r0mant) [inline lib/web/saml.go]: access denied

File: lib/web/saml.go [hunk 10]
```diff
@@ -90,9 +102,15 @@ func (m *Handler) samlACS(w http.ResponseWriter, r *http.Request, p httprouter.P
 		http.Redirect(w, r, pathToError.String(), http.StatusFound)
 		return nil, nil
 	}
+
 	// if we created web session, set session cookie and redirect to original url
 	if response.Req.CreateWebSession {
 		log.Debugf("redirecting to web browser")
+		err = csrf.VerifyToken(response.Req.State.CSRFToken, r)
+		if err != nil {
+			return nil, trace.AccessDenied("unable to verify CSRF token", err)
```
Comment (r0mant) [inline lib/web/saml.go]: "access denied"

File: lib/web/ui/server.go [hunk 11]
```diff
@@ -44,26 +44,43 @@ type Server struct {
 	Labels []Label `json:"tags"`
 }
 
+type sortedLabels []Label
+
+func (s sortedLabels) Len() int {
```
Comment (r0mant) [inline lib/web/ui/server.go]: Can you add godocs on these methods? (`Len`, `Less`, `Swap`)

**Comment by alex-kovoy** [comment]
retest this please

File: lib/httplib/csrf/csrf.go [hunk 12]
```diff
@@ -40,72 +40,86 @@ const (
...
+// VerifyHTTPHeader checks if HTTP header value matches the cookie.
+func VerifyHTTPHeader(r *http.Request) error {
+	token := r.Header.Get(HeaderName)
+	if len(token) == 0 {
+		return trace.BadParameter("cannot retrieve CSRF token from HTTP header %q", HeaderName)
 	}
 
-	save(encodedToken, w)
-	return encodedToken, nil
+	err := VerifyToken(token, r)
```
Comment (kontsevoy) [inline lib/httplib/csrf/csrf.go]: this block below can be one-liner: `return trace.Wrap(VerifyToken(token, r))`. No need to check for nils.

File: lib/httplib/httplib.go [hunk 13]
```diff
@@ -69,6 +72,20 @@ func MakeStdHandler(fn StdHandlerFunc) http.HandlerFunc {
 	}
 }
 
+// WithCSRFProtection ensures that request to unauthenticated API is checked against CSRF attacks
+func WithCSRFProtection(fn HandlerFunc) httprouter.Handle {
+	hanlderFn := MakeHandler(fn)
+	return func(w http.ResponseWriter, r *http.Request, p httprouter.Params) {
+		err := csrf.VerifyHTTPHeader(r)
```
Comment (kontsevoy) [inline lib/httplib/httplib.go]: so... we have `err` here which is supposed to be useful and informative (it is returned from a higher level funcion `VerifyHTTPHeader()`. but `err` is not used, and instead is wrapped (below) into another two different (again, why different?) error messages. why not have  `VerifyHTTPHeader()` return something which is good enough to pass directly to the log and return on the wire?

File: lib/web/ui/server.go [hunk 14]
```diff
@@ -44,26 +44,44 @@ type Server struct {
 	Labels []Label `json:"tags"`
 }
 
+// sortedLabels is a sort wrapper that sorts labels by name
+type sortedLabels []Label
```
Comment (kontsevoy) [inline lib/web/ui/server.go]: maybe we should get rid of `Label` type definition and just use `string`. It prevents you from using `sort.Strings()` and you have to create these tiny methods for no reason.

**Review by kontsevoy — COMMENTED** [review]
left some thoughts for you and Sasha to consider.

**Review by klizhentas — APPROVED** [review]

File: lib/web/ui/server.go (hunk 14 above)
Comment (klizhentas) [reply lib/web/ui/server.go]: in the future, let's keep PR per feature/bug please @alex-kovoy

**Pull request merged**